### PR TITLE
eth_call_v2

### PR DESCRIPTION
**Pull request by zjb0807**
Added `eth_call_v2` version, now both calls are supported.
We can remove `eth_call` in a later runtime upgrade, it is no longer supported.

### CODE REVIEW

File: primitives/src/evm.rs [hunk 1]
```diff
@@ -222,6 +230,58 @@ impl TryFrom<CurrencyId> for EvmAddress {
...
+
+	// percentage
+	let tip_number = gas_price.checked_div(ONE_GWEI)?.checked_sub(100)?;
+	if !tip_number.is_zero() {
+		actual_gas_price = gas_price.checked_sub(tip_number.checked_mul(ONE_GWEI)?)?;
+		tip = actual_gas_price
+			.checked_mul(gas_limit)?
+			.checked_mul(tip_number)?
+			.checked_div(100)? // percentage
+			.checked_div(1_000_000)? // ACA decimail is 12, ETH decimail is 18
```
Comment (zjb0807) [inline primitives/src/evm.rs:251]: @shunjizhan I added this line, you should do the same in bodhi.js

File: primitives/src/evm.rs [hunk 2]
```diff
@@ -222,6 +230,58 @@ impl TryFrom<CurrencyId> for EvmAddress {
...
+			.into();
+	}
+
+	// valid_until max is 999_999_999
+	let valid_until: u32 = Into::<u128>::into(actual_gas_price)
+		.checked_sub(tx_fee_per_gas)?
+		.try_into()
+		.ok()?;
```
Comment (zjb0807) [inline primitives/src/evm.rs:259]: `gasPrice = 100 Gwei = 100_000_000_000`,  so valid_until max is `999_999_999`. 
`BlockNumber = u32` , `u32::MAX; // 4_294_967_295u32`.  Decimal is not enough here, but it should be fine.

Comment (xlc) [reply primitives/src/evm.rs:259]: with 6s block time, the world is going to end in 190 years? 💀

Comment (shunjizhan) [reply primitives/src/evm.rs:259]: :joy:

Comment (zjb0807) [reply primitives/src/evm.rs]: Now use 10 Gwei(`10_000_000_000`) to store valid_until, and `valid_until = min(u32::MAX, valid_until)`

File: primitives/src/evm.rs [hunk 3]
```diff
@@ -238,25 +238,25 @@ pub fn decode_gas_price(gas_price: u64, gas_limit: u64, tx_fee_per_gas: u128) ->
 
 	let mut tip: u128 = 0;
 	let mut actual_gas_price = gas_price;
-	const ONE_GWEI: u64 = 1_000_000_000u64;
+	const TEN_GWEI: u64 = 10_000_000_000u64;
 
-	// percentage
-	let tip_number = gas_price.checked_div(ONE_GWEI)?.checked_sub(100)?;
+	// tip = 10% * tip_number
+	let tip_number = gas_price.checked_div(TEN_GWEI)?.checked_sub(10)?;
```
Comment (shunjizhan) [inline primitives/src/evm.rs:244]: should `TEN_GWEI` be 1000 Gwei instead?

File: primitives/src/evm.rs [hunk 4]
```diff
@@ -238,25 +238,25 @@ pub fn decode_gas_price(gas_price: u64, gas_limit: u64, tx_fee_per_gas: u128) ->
...
-			.checked_div(100)? // percentage
+			.checked_div(10)? // percentage
 			.checked_div(1_000_000)? // ACA decimal is 12, ETH decimal is 18
 			.into();
 	}
 
-	// valid_until max is 999_999_999
+	// valid_until max is u32::MAX.
 	let valid_until: u32 = Into::<u128>::into(actual_gas_price)
 		.checked_sub(tx_fee_per_gas)?
```
Comment (shunjizhan) [inline primitives/src/evm.rs:257]: should this be `.checked_sub(1000 Gwei)`?

**Review by shunjizhan — APPROVED** [review]
LGTM

File: primitives/src/evm.rs (hunk 2 above)
Comment (zjb0807) [reply primitives/src/evm.rs]: @xlc please take a look

File: primitives/src/evm.rs [hunk 5]
```diff
@@ -222,6 +230,58 @@ impl TryFrom<CurrencyId> for EvmAddress {
...
+			.checked_mul(tip_number)?
+			.checked_div(10)? // percentage
+			.checked_div(1_000_000)? // ACA decimal is 12, ETH decimal is 18
+			.into();
+	}
+
+	// valid_until max is u32::MAX.
+	let valid_until: u32 = Into::<u128>::into(actual_gas_price)
+		.checked_sub(tx_fee_per_gas)?
+		.try_into()
```
Comment (xlc) [inline primitives/src/evm.rs:258]: this is `saturated_into`

**Review by xlc — DISMISSED** [review]

**Review by xlc — APPROVED** [review]

**Pull request merged**